### PR TITLE
No database dependency during inference/serving

### LINKED ISSUE #316: No database dependency during inference/serving
**Is your feature request related to a problem? Please describe.**

I think a Fonduer-based app lifecycle has three phases: development, training, and serving.
During development, you may have many iterations of re-definition of mention/candidate subclasses, labeling functions, which involve re-parsing, re-extraction, re-labeling, etc.
I understand that it is important to persist intermediate results to a database to save time.
Depending on the size of training dataset, you may need to persist intermediate results during training too.

On the other hand, there is no need to persist intermediate results to a database during inference/serving.
The dependency on database (postgres) makes a Fonduer-based app less portable and less scalable.

**Describe the solution you'd like**

I'd like no database dependency during inference/serving.

**Describe alternatives you've considered**

sqlite is easier to install than postgres, but still there is no need to persist intermediate results to a database.

**Additional context**

Related to #137.

### PR DESCRIPTION
**Pull request by HiromuHota**
This will close #316 once it's finished.

### CODE REVIEW

**Comment by HiromuHota** [comment]
Although it needs some code cleanup, it's now functioning.
As can be seen in https://github.com/HiromuHota/fonduer-mlflow/tree/fix/316, a database is not needed during inference.
I'll do some code cleaning and make it ready for review shortly.

**Comment by HiromuHota** [comment]
The number of commits was reduced to 18, but still a lot to review.
I'll be splitting them into smaller PRs so that it is easier to review.

File: src/fonduer/parser/parser.py [hunk 1]
```diff
@@ -249,13 +249,13 @@ def __init__(
 
     def apply(  # type: ignore
         self, document: Document, pdf_path: Optional[str] = None, **kwargs: Any
-    ) -> Iterator[Sentence]:
+    ) -> Document:
         # The document is the Document model
         text = document.text
 
         # Only return sentences, if no exceptions occur during parsing
```
Comment (lukehsiao) [inline src/fonduer/parser/parser.py]: Let's update this comment, too. Since we are returning a document now, not sentences.

**Comment by lukehsiao** [comment]
I know this is unrelated to this specific PR, but I wonder if there is a way to eliminate the ~1k warnings we get in test_e2e on travis.

File: src/fonduer/parser/parser.py (hunk 1 above)
Comment (HiromuHota) [reply src/fonduer/parser/parser.py]: Good catch. I update it by simply removing the comment and adding a docstring.

**Comment by HiromuHota** [comment]
If I remember correctly, the ~1k warnings come from Snorkel where it uses deprecated APIs of PyTorch.

File: src/fonduer/candidates/candidates.py [hunk 2]
```diff
@@ -283,6 +280,17 @@ def apply(  # type: ignore
                     for mention in candidate_class.mentions
                 ]
             )
+            # Construct a set of stable ids of candidates.
+            set_of_stable_ids = set()
+            if hasattr(doc, candidate_class.__tablename__ + "s"):
```
Comment (SenWu) [inline src/fonduer/candidates/candidates.py]: Can we have a documentation here to mention why `+ "s"`?

File: src/fonduer/candidates/candidates.py [hunk 3]
```diff
@@ -283,6 +280,17 @@ def apply(  # type: ignore
...
+                set_of_stable_ids.update(
+                    set(
+                        [
+                            tuple(m.context.get_stable_id() for m in c)
+                            for c in getattr(doc, candidate_class.__tablename__ + "s")
```
Comment (SenWu) [inline src/fonduer/candidates/candidates.py]: Same here.

File: src/fonduer/candidates/candidates.py [hunk 4]
```diff
@@ -314,16 +322,18 @@ def apply(  # type: ignore
...
-                if not clear:
-                    q = select([candidate_class.id])
-                    for key, value in list(candidate_args.items()):
-                        q = q.where(getattr(candidate_class, key) == value)
-                    candidate_id = self.session.execute(q).first()
-                    if candidate_id is not None:
-                        continue
+                    candidate_args[arg_name] = cand[j][1]
+
+                stable_ids = tuple(
```
Comment (SenWu) [inline src/fonduer/candidates/candidates.py]: Can we have a global func to construct stable id? It's a little bit messy to have stable id constructor everywhere.

File: src/fonduer/candidates/mentions.py [hunk 5]
```diff
@@ -577,46 +563,61 @@ def __init__(
...
-    def apply(  # type: ignore
-        self, doc: Document, clear: bool, **kwargs: Any
-    ) -> Iterator[Mention]:
+    def apply(self, doc: Document, **kwargs: Any) -> Document:
         """Extract mentions from the given Document.
 
         :param doc: A document to process.
-        :param clear: Whether or not to clear the existing database entries.
         """
+        # Construct a dict of stable ids of contexts.
```
Comment (SenWu) [inline src/fonduer/candidates/mentions.py]: Is it possible to move this to the utils?

File: src/fonduer/candidates/candidates.py (hunk 4 above)
Comment (HiromuHota) [reply src/fonduer/candidates/candidates.py]: I moved these stable_id constructors to `fonduer.utils.utils`.

File: src/fonduer/candidates/candidates.py (hunk 2 above)
Comment (HiromuHota) [reply src/fonduer/candidates/candidates.py]: I added a comment to explain the reason.
See https://github.com/HazyResearch/fonduer/pull/368/files#diff-dd77b82c8178c68913da115c82628c3dR56.

File: src/fonduer/candidates/candidates.py (hunk 3 above)
Comment (HiromuHota) [reply src/fonduer/candidates/candidates.py]: See https://github.com/HazyResearch/fonduer/pull/368/files#diff-dd77b82c8178c68913da115c82628c3dR56.

**Pull request merged**